File: BaekJoon/C++/14889_스타트와링크.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace std;

int N, mymin = 3333333;
int S[20][20] = {};
bool selected[20] = {};

void startlink(int num, int cnt){
    vector<int> start_team;
    vector<int> link_team;
    int start_score = 0;
    int link_score = 0;
    
    if(cnt == N / 2){
        for(int i = 0; i < N; i++){
            if(selected[i] == true)
                start_team.push_back(i);
            else
                link_team.push_back(i);
        }
        
        for(int i = 0; i < (N/2); i++){
            for(int j = 0; j < (N/2); j++){
                start_score += S[start_team[i]][start_team[j]];
                link_score += S[link_team[i]][link_team[j]];
            }
        }
        
        if(abs(start_score - link_score) < mymin)
            mymin = abs(start_score - link_score);
        
        return;
    }
    
    for(int i = num; i < N; i++){
        if(!selected[i]){
            selected[i] = true;
            startlink(i, cnt + 1);
            selected[i] = false;
        }
    }
    
}

int main() {
    cin >> N;
    for(int i = 0; i < N; i++){
        for(int j = 0; j < N; j++){
            scanf("%d", &S[i][j]);
        }
    }
    startlink(0, 0);
    printf("%d", mymin);
    
    return 0;
}

File: BaekJoon/Python/14889_스타트와링크.py
def func(idx, cnt):
    global answer

    if cnt == N // 2:
        start_score, link_score = 0, 0
        for i in range(N):
            for j in range(N):
                # 선택된 번호는 스타트팀, 선택되지 않은 번호는 링크팀
                if selected[i] and selected[j]:
                    start_score += S[i][j]
                elif not selected[i] and not selected[j]:
                    link_score += S[i][j]
        answer = min(answer, abs(start_score - link_score))
        return

    for i in range(idx, N):
        if not selected[i]:
            selected[i] = True
            func(i+1, cnt+1)
            selected[i] = False


N = int(input())
S = [list(map(int, input().split())) for _ in range(N)]
selected = [False for _ in range(N)]
answer = 1e9

func(0, 0)
print(answer)
